### PR TITLE
implement masked_arith_op to de-duplicate ops code

### PR DESCRIPTION
**Pull request by jbrockmendel**
ATM _arith_method_SERIES and _arith_method_FRAME have really similar fallback code.  This de-duplicates those.

- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: pandas/core/ops.py [hunk 1]
```diff
@@ -788,6 +789,55 @@ def mask_cmp_op(x, y, op, allowed_types):
...
+    only non-null elements.
+
+    Parameters
+    ----------
+    x : np.ndarray
+    y : np.ndarray, Series, Index
+    op : binary operator
+    """
+    xrav = x.ravel()  # 2D compat
+    assert isinstance(x, (np.ndarray, ABCSeries)), type(x)
```
Comment (gfyoung) [inline pandas/core/ops.py]: Can we provide a more informative error message than just the `type` of `x`?

Comment (gfyoung) [reply pandas/core/ops.py]: This comment BTW applies to all of the `assert` statements that you added.

**Comment by jreback** [comment]
looks ok, if you can comment a bit more, this consolidates a gnarly bit of code, which is full of special cases. and rebase.

**Comment by jreback** [comment]
thanks

**Pull request merged**